### ISSUE TITLE
provision-models emits an fp32 gender model (+190 MB, ~+500 MiB VRAM); restore the fp16 export

### ISSUE THREAD

**Issue by attevon-admin** [attevon-llc/diar-native#6]
## Summary

`provision-models` reproduces **23 of 24** files in `models_folded/` exactly. The one divergence is `gender-wav2vec2.onnx`, which comes out **fp32 (379 MB)** instead of the shipped **fp16 (189 MB)**.

This is a size/VRAM regression, **not** a correctness bug — measured below — but it should be closed before the fp32 build becomes the de facto artifact.

## Verification that provisioning is otherwise exact

Cold-cache end-to-end run (empty models dir, empty HF cache, real gated download):

```
Provisioned /models (fast set, 673 MB) in 119.5s. Smoke test passed.
```

`scripts/compare_model_sets.py models_folded <new>`:
- All 6 `plda_*.npy` + `wespeaker-voxceleb-resnet34.min_num_samples.txt` — **byte-identical**
- All **15** diarization ONNX graphs — op histograms identical, **every initializer tensor identical**
- Only `gender-wav2vec2.onnx` fails Tier A (size) and Tier B (op histogram)

## Gender model equivalence — measured, 6 seeded inputs, CPU EP

```
old in [('input_values', ['batch','samples'], 'tensor(float)')]
new in [('input_values', ['batch','samples'], 'tensor(float)')]
  2s maxdiff=0.004719 label 1==1 OK
  3s maxdiff=0.002193 label 1==1 OK
  4s maxdiff=0.005014 label 1==1 OK
  5s maxdiff=0.004001 label 1==1 OK
  6s maxdiff=0.006 -> worst overall 0.00501442
RESULT label_agreement=6/6 worst_logit_diff=0.00501442
```

Identical I/O signature (`input_values` -> `logits`, the exact names `crates/diar-core/src/gender.rs` requires). Divergence is consistent with fp16-vs-fp32 rounding.

The op-histogram delta (`IsNaN` 0->12, `Where` 0->25, `Expand` 0->1, `Range` 0->1, `Constant` 296->346) reflects a different tracer emitting NaN guards, not different weights.

## Root cause

fp16 conversion fails under **torch 2.13**; the shipped fp16 artifact was produced under torch 2.11.

```
fp16 conversion unavailable ([ONNXRuntimeError] : 1 : FAIL :
Load model from gender-wav2vec2.onnx.fp16 failed: Type Error: Type (tensor(float16)...
```

`onnxconverter_common.float16` also emits many `will be truncated to 1e-07` warnings on subnormals.

The exporter correctly detects the failure and falls back to fp32 rather than shipping a broken graph — the fail-safe behaved properly.

## Cost of the fallback

- **+190 MB** on disk (models dir 673 MB vs ~470 MB reference)
- **~+500 MiB VRAM** at runtime (RESULTS §7.18)
- Gender is already 74% of the ~620 MB RSS a resident second engine costs (RESULTS §7.34), so this compounds on multi-device deployments

## Options

1. Pin torch 2.11 in `scripts/provision/requirements.txt` for the gender step only.
2. Fix the fp16 conversion under torch 2.13 (likely a `keep_io_types` / blocked-op-list adjustment in `onnxconverter_common.float16`).
3. Convert fp16 from the fp32 ONNX as a separate post-step with an explicit op blocklist, then validate with the numeric check above.

Option 3 is probably the most durable — it decouples precision from the torch version entirely.

## Acceptance

- [ ] `provision-models` emits an fp16 `gender-wav2vec2.onnx` at ~189 MB.
- [ ] `compare_model_sets.py models_folded <new>` reports **0 failures**.
- [ ] Label agreement vs the shipped model stays 6/6 on the seeded probe, with `max|delta logit|` within fp16 tolerance.
- [ ] `--skip-gender` remains available for operators who do not need it.

## Interim state

The fp32 model is correct and safe to use. `docs/INSTALL_NATIVE.md` should state the provisioned set is ~673 MB (not ~470 MB) until this is fixed.

**Comment by attevon-admin** [comment]
Fixed in 0ac4e63. Full evidence in `validation/RESULTS.md` §7.39.

**Root cause — not a converter limitation.** torch 2.13 emits two semantically invisible `Cast` nodes that torch 2.11 did not: `/inner/wav2vec2/encoder/Cast_1` = `Cast(to=FLOAT)` on an already-fp32 `Transpose` output feeding `encoder/Add`, and `/inner/wav2vec2/encoder/Cast` = `Cast(to=INT64)` on an already-int64 value. `onnxconverter_common` treats every `Cast` as an authoritative precision boundary — casts are the mechanism it inserts itself to bridge fp16/fp32 under `keep_io_types` — so it retypes the value to fp16 and leaves `to=FLOAT` alone. ORT then rejects the contradiction.

This also resolves the confusing symptom pair in §7.36. Both errors are one bug:

```
baseline             FAIL  Type (tensor(float16)) of output arg (.../encoder/Cast_1_output_0)
                           of node (.../encoder/Cast_1) ... expected tensor(float)
disable_shape_infer  FAIL  Type parameter (T) of Optype (Add) bound to different types
                           (tensor(float16) and tensor(float)) in node (.../encoder/Add)
```

`/inner/wav2vec2/encoder/Add` is exactly the node `Cast_1` feeds — "moves the error from a Cast to an Add" was the same node all along.

**Fix.** `_elide_identity_casts` drops any `Cast` whose `to` already equals its input's inferred element type, before conversion (never one feeding a graph output). No torch pin needed.

**Gates, now run by the exporter on every run** over 6 seeded `do_normalize` clips at 16k/24k/32k/48k/64k/80k samples (bracketing `gender.rs`'s `MIN_SAMPLES` .. `DIAR_GENDER_MAX_SECONDS` range):

| gate | bar | measured |
|---|---|---|
| fp32 ONNX vs torch | < 1e-4 | 5.36e-06 |
| elision is a no-op | bitwise equal, 6/6 | 6/6, max abs diff 0.000e+00 |
| fp16 argmax invariance | 6/6 labels | 6/6, max abs Δlogit 1.06e-02, max abs Δp 4.86e-04 (bar 0.05) |

**Result.** `gender_precision: "fp16"` on the pinned torch 2.13.0 stack. **378,529,501 → 189,488,828 bytes (−50.0%)**, 213/213 FLOAT16 initializers with fp32 I/O — matching the shipped torch-2.11 artifact (189,431,659 B) on every load-bearing property. It is 57,169 B larger because torch 2.13 emits extra attention NaN-clamp/mask machinery (1084 vs 966 nodes); that is present in the fp32 export too and does not touch the weights.

**Verified end to end under ort `=2.0.0-rc.12`**, two containers differing only by a bind-mount of the gender model, `karpathy_10m.wav`, `gender: true`:

- `rttm` sha256 `edcc5d27…b4ef` — identical
- `segments` (92), `exclusive_segments` (88), `centroids`, `num_speakers`: equal
- gender **2/2 labels agree**: SPEAKER_00 female 0.79675186 vs 0.79675025 (Δ 1.6e-06), SPEAKER_01 male 0.9991439 vs 0.9991439 — and both match §7.16's recorded `female 0.797 / male 0.999` for this clip.

**Rejected:** pinning torch 2.11 for provisioning (freezes the export env against the rest of the pinned stack to dodge two no-op nodes, and OpenTranscribe's backend image — the primary provisioning route — does not get to choose its torch version); `op_block_list=['Cast']` (blocking the op the converter uses as its own precision boundary is what causes the mismatch).

Cost: ~11 s added to provisioning (one shape-inference pass over a 378 MB graph, 2 session loads, 12 CPU inferences) against §7.36's 119.5 s cold run — a one-time cost for 189 MB disk and ~500 MiB VRAM permanently. That timing is load-contaminated and order-of-magnitude only.

Not closed by this, and tracked separately: `/healthz` `models_gender` is still a bool of file existence and carries no precision, so a *serving* directory still can't be asked which precision it has without reading the marker off disk. (The serde drop itself is fixed — `gender_precision` now reaches `Toolchain` and the provisioning report.)
